### PR TITLE
Refactor code and introduce new standards

### LINKED ISSUE #17: Better python package naming
`api` is too broad of a name, should be something like `hmtescrow`

### LINKED ISSUE #18: Simple api usage documentation
Should add some basic usage of the package into the readme of this repo

### LINKED ISSUE #62: Break out manifest base models into their own repo / package
Currently the way that `basemodels` lives in hmt-escrow is not ideal as in order to use it in other places (ie hmt-servers) you need to basically rebuild and pull in all of the contracts, hmt-escrow content, and dependent libs .. 

It would be much cleaner and simple to upgrade a manifest if it lived in its own repo, versioned in the same way that hmt-escrow is 

cc: @rbval @posix4e

### LINKED ISSUE #66: Confusing function names in escrow and API, some dead code
We have gotten a bit sloppy with our naming conventions in this package. 

1. `payout` and `bulk_payout` take in **final** results uri+hash. They should make that clear.
2. store_results relates only to **intermediate** results. It should be renamed.
3. _partial_payout appears to be unused now that bulk payouts are implemented, and may be eliminated.

### LINKED ISSUE #67: payouts in Python API need better return handling
These functions need to return something related to success or failure, rather than undefined.

### LINKED ISSUE #68: Clean code, improve code quality, lift coding standards
1. Double check for mypy types
2. Add necessary comments
3. Refactor out some confusing variable or function names

### LINKED ISSUE #71: Move existing tests to doctests

### PR DESCRIPTION
**Pull request by rbval**
Closes #17 
Closes #18
Closes #62
Closes #66 
Closes #67
Closes #68 
Closes #71

As we are moving nearer to mainnet and we want to achieve some certain of maturity in our codebase, I started looking at our code and made a lot of changes that should now make the codebase clearer.
These changes make the current and previous version non-backward compatible. More of the examples can be found in the updated README.

In essence here's the bird's eye view to changes:

Python API:
Credentials containing the address and the private key of the gas payer need to be provided upon instantiation. Instantiating a Job class can be done three different ways:
- Launch a fresh Job and a fresh factory
- Launch a fresh Job with an existing factory
- Access an existing Job with an existing factory and escrow address

Job:
- `Contract` renamed to `Job` class as `WContract` was confusing
- `api` package renamed to `hmt_escrow` package
- `deploy` has been renamed to `launch`
- `setup` now ties old `fund` and `setup` together
- `payout` and `partial_payout` have been removed
- `get_contract_from_address` has been removed
- `get_job` has been removed
- `get_job_from_address` has been removed as `get_escrow` from `eth_bridge` can be used directly
-  extranous `_sol` endings have been removed
- `initialize` is handled by `__init__` and `_initialize_job`
- Add `_access_job` used by `__init__` to initialize the Job based on an existing escrow contract's manifest
- Add `_counter`, `_last_address` and `_create_escrow` helper methods for factory logic to make code clearer and more simple
- Rename helper methods regarding intermediate / manifest / final results clearer
- Adjust `complete` logic to do write and read separately
- Use the status as Enum to access contract status
- All methods updating contract's state call eth_bridge's high-order function `handle_transaction`

Eth Bridge:
- Infuse `send_and_sign_transaction` and `wait_on_transaction` to a high-order function called `handle_transaction` that can be used to update a contract's state

Storage:
- Remove unneeded private keys in encrypting and uploading to IPFS
- Handle Exceptions regarding IPFS connections

General:
- Add and modify all docstrings to follow the Google's PEP-484 based standard
- Add extensive and explicit typing in all methods possible
- Add doctests to all relevant functions and methods.

Solidity:
- Remove `payOut` and `partialPayout` logic
- Rename confusing manifest hash names to clearer ones
- Change `complete` to focus on only updating the state as transactions in Ethereum don't have return values
- Add a mapping to check for escrow addresses deployed by a factory.

### CODE REVIEW

**Comment by rbval** [comment]
I guess the issue @gaieges is that other changes would be regarding basemodels, let's decouple that immediately so we can follow SemVer properly?

Was just browsing some python doc generators, I think I'll create another PR for that.

File: hmt_escrow/storage.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+LOG = logging.getLogger("api.storage")
+
+if not os.getenv("IPFS_DISABLE"):
+    _host = os.getenv("IPFS_HOSTNAME", 'localhost')
+    _port = int(os.getenv("IPFS_TCP_PORT", '5001'))
+    try:
+        API = ipfsapi.connect(_host, _port)
+    except Exception as e:
+        raise e
```
[the remark this answers is not in the document]
Comment (posix4e) [reply hmt_escrow/storage.py]: i’m down to turning this into a warn

**Comment by posix4e** [comment]
glad this worked out lgtm

**Comment by rbval** [comment]
Would appreciate @durch and @gaieges 's comments before merging. We can ticket those up if not fixing in this PR.

File: setup.py [hunk 2]
```diff
@@ -16,5 +16,5 @@
     packages=setuptools.find_packages(),
     install_requires=[
         "ipfsapi==0.4.3", "py-evm==0.2.0a34", "py-solc==3.2.0",
-        "schematics==2.1.0", "web3==4.8.2", "yapf==0.25.0"
+        "schematics==2.1.0", "web3==4.8.2", "yapf==0.25.0", "mypy==0.670"
```
Comment (gaieges) [inline setup.py]: You can probably pull schematics as that now should be in `hmt-basemodels`

File: hmt_escrow/job.py [hunk 3]
```diff
@@ -0,0 +1,1021 @@
...
+from eth_keys import keys
+from eth_utils import decode_hex
+
+from eth_bridge import get_hmtoken, get_contract_interface, get_escrow, get_factory, deploy_factory, get_w3, handle_transaction
+from storage import download, upload
+from basemodels import Manifest
+
+GAS_LIMIT = int(os.getenv("GAS_LIMIT", 4712388))
+
+LOG = logging.getLogger("hmt_escrow.job")
```
Comment (gaieges) [inline hmt_escrow/job.py]: nice logging hygiene

File: hmt_escrow/job.py [hunk 4]
```diff
@@ -0,0 +1,1021 @@
...
+        manifest_url (str): the location of the serialized manifest in IPFS.
+        manifest_hash (str): SHA-1 hashed version of the serialized manifest.
+
+    """
+
+    def __init__(self,
+                 credentials: Dict[str, str],
+                 escrow_manifest: Manifest = None,
+                 factory_addr: str = None,
+                 escrow_addr: str = None):
```
Comment (gaieges) [inline hmt_escrow/job.py]: This simple api is way cleaner and easy to understand.  Really like how you can point it an existing factory/ address or have it create a new one that allows you to operate off of that.

**Comment by rbval** [comment]
> All really great! Heroic effort here.. well done. Very helpful summary in the PR too, could be used in a changelog as well. Quick little notes, nothing major.
> 
> Majority of our tests here are now doctests huh? That requires a ganache running to run them?

Yap, all our tests are in doctests. It requires ganache running, yes. It takess approx 35 secs total from spinning of `bin/ci` / `bin/test` to run the python tests and 1 min total with the js tests.

File: setup.py (hunk 2 above)
Comment (rbval) [reply setup.py]: Good point, fixed!

**Pull request merged**